rfcs/20200804-primitive-fallback/README.md

```diff
@@ -0,0 +1,172 @@
+there is no error handling mechanism for JIT kernels on GPU, and we may revisit
+a similar issue for GPU primivies in the future.
+
+## Background
+
+Primitives in oneDNN internally have multiple implementations that (i) are
+optimized for a particular Instruction Set Architecture (ISA), and (ii) can
+handle a particular set of user parameters (e.g. attributes, tensor shapes,
+memory layout etc). All (or most) primitives contain implementations that use
+Just-In-Time (JIT) code generation in order to further improve performance.
+Currently, generation of JIT code (or kernel) is a two-step process:
+1. An implementation, that satisfies the above criteria, is being looked for
+   during primitive descriptor construction (e.g.
+   *convolution_forward::primitive_desc()*);
+2. If the implementation above is based on a JIT kernel, the corresponding code
+   is being generated during construction of the primitive (e.g.
+   *convolution_forward()*)
+
+Currently, if an implementation has been successfully found, it is expected that
+it will also successfully generate the kernel and execute the generated code.
+However, occasionally, a JIT implementation may throw an exception during
+generation [1], which may be a simple logical error or something much less
+trivial to fix.
+
+## Proposal
+
+This document presents some options to deal with inability of a primitive to be
+constructed having created a valid primitive descriptor implementation.
+
+The options below are listed in the order of increased amount of anticipated
+changes to support primitive fallback.
+
+Additional options are welcome to be added to the list.
+
+### Option 1: leave as is
+
+For this option, the user will take advantage of the existing
+`primitive_desc::next_impl()` API to be able to deploy a different
+implementation if the first one fails. In this case, the user is enforced to
+introduce some control flow, something like
+~~~cpp
+
+auto pd = convolution_forward::primitive_desc(opd, attr, eng);
+convolution_forward prim;
+
+do {
+    try {
+        prim = convolution_forward(pd);
+    }
+    catch (dnnl::error &e) {...}
+} while (prim == convolution_forward() && pd.next_impl());
+~~~
+
+In order to ensure that the issue described above will be taken care of by the
+user, the documentation will be updated accordingly.
+
+Pros: 
+1. Minimal changes required in documentation;
+
+Cons:
+1. The user is responsible for ensuring that a valid primitive implementation
+   exists.
+
+### Option 2: allow construction of an empty primitive
+
+For this option, we attempt to simplify the control flow in [Option
+1](#option-1-leave-as-is) by allowing construction of an empty primitive without
+throwing an exception using the `allow_empty` flag:
+
+~~~cpp
+primitive::primitive(const_dnnl_primitive_desc_t c_pd, bool allow_empty = false)
+~~~
+
+In this case, the control flow can be simplified to:
+
+~~~cpp
+
+auto pd = convolution_forward::primitive_desc(opd, attr, eng);
```

Comment:
I think it would be better to introduce `bool primitive::empty()` given that we allow to create an empty primitive.

Reply:
Turns out we have `operator bool()` defined, but I guess we can introduce an extra API for ease of use in certain cases.

Reply:
I think the `operator bool()'s would be sufficient. Resolving this.